sklearn/calibration.py

```diff
@@ -887,6 +909,149 @@ def loss_grad(AB):
+    labels = labels.astype(int)
+    max_logits = np.reshape(np.max(logits, axis=1), (-1, 1))
+
+    def negative_log_likelihood(beta, logits, labels, max_logits):
+        """Compute the negative log likelihood loss and its derivative
+            with respect  to temperature.
+
+        Parameters
+        ----------
+        beta : float
+            The current inverse temperature value during optimisation.
+
+        logits : ndarray of shape (n_samples, n_classes)
+            The decision function or predict proba for the samples.
+
+        labels : ndarray of shape (n_samples, 1) or (n_samples, n_classes)
+            True labels for the samples. For binary classification, it
+            is given by a binary column vector. For multi-class classification,
+            it is given by one-hot vectors.
+
+        max_logits : ndarray of shape (n_samples, 1)
+            Maximum value in each row of logits
+
+        Returns
+        -------
+        negative_log_likelihood_loss : float
+            The negative log likelihood loss.
+
+        negative_log_likelihood_loss_derivative : float
+            The derivative of the negative log likelihood loss with respect to
+            the inverse temperature.
+        """
+
+        # Select the logit and the softmax output of the correct class
+        softmax_output = softmax(beta * logits)
+
+        # multi-class classification
+        if (labels.ndim == 2) and (labels.shape[1] > 1):
+            correct_class_probas = np.sum(softmax_output * labels, axis=1)
+            correct_class_logits = np.sum(logits * labels, axis=1)
+
+        # binary classification
+        elif (labels.ndim == 2) and (labels.shape[1] == 1):
+            correct_class_probas = np.take_along_axis(
+                softmax_output, labels, axis=1
+            ).ravel()
+
+            correct_class_logits = np.take_along_axis(logits, labels, axis=1).ravel()
+
+        # Compute the losses
+        losses = np.log(correct_class_probas)
+
+        # Derivatives with respect to beta
+        numerator = np.sum(logits * np.exp(beta * (logits - max_logits)), axis=1)
+
+        denominator = np.sum(np.exp(beta * (logits - max_logits)), axis=1)
+
+        derivatives = correct_class_logits - numerator / denominator
+
+        # Apply sample weight
+        if sample_weight is not None:
+            losses *= sample_weight
+            derivatives *= sample_weight
+
+        return -np.sum(losses), -np.sum(derivatives)
+
+    beta_minimizer = minimize(
+        negative_log_likelihood,
+        np.array([beta_0]),
+        args=(logits, labels, max_logits),
+        method="L-BFGS-B",
+        bounds=[(1e-8, None)],
+        jac=True,
+        options={
+            "gtol": 1e-6,
+            "ftol": 64 * np.finfo(float).eps,
+        },
+    )
+
+    return beta_minimizer.x[0]
+
+
 class _SigmoidCalibration(RegressorMixin, BaseEstimator):
     """Sigmoid regression model.
```

Comment:
We should override `__sklearn_tags__` to set the same input tags as `IsotonicRegression`:

https://github.com/scikit-learn/scikit-learn/blob/af41352a8c8ac7409d86860ab192742d709c6275/sklearn/isotonic.py#L513-L517

See:

https://github.com/scikit-learn/scikit-learn/blob/af41352a8c8ac7409d86860ab192742d709c6275/sklearn/utils/_tags.py#L20-L25